x/virtualgroup/keeper/keeper.go

```diff
@@ -0,0 +1,680 @@
+	"fmt"
+
+	"cosmossdk.io/errors"
+	"cosmossdk.io/math"
+	"github.com/cometbft/cometbft/libs/log"
+	"github.com/cosmos/cosmos-sdk/codec"
+	"github.com/cosmos/cosmos-sdk/store/prefix"
+	storetypes "github.com/cosmos/cosmos-sdk/store/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/address"
+	"github.com/prysmaticlabs/prysm/crypto/bls"
+
+	"github.com/bnb-chain/greenfield/internal/sequence"
+	types2 "github.com/bnb-chain/greenfield/types"
+	sptypes "github.com/bnb-chain/greenfield/x/sp/types"
+	storagetypes "github.com/bnb-chain/greenfield/x/storage/types"
+	"github.com/bnb-chain/greenfield/x/virtualgroup/types"
+)
+
+type (
+	Keeper struct {
+		cdc       codec.BinaryCodec
+		storeKey  storetypes.StoreKey
+		tStoreKey storetypes.StoreKey
+		authority string
+
+		// Keepers
+		spKeeper      types.SpKeeper
+		accountKeeper types.AccountKeeper
+		bankKeeper    types.BankKeeper
+		paymentKeeper types.PaymentKeeper
+		// sequence
+		lvgSequence       sequence.Sequence[uint32]
+		gvgSequence       sequence.Sequence[uint32]
+		gvgFamilySequence sequence.Sequence[uint32]
+	}
+)
+
+func NewKeeper(
+	cdc codec.BinaryCodec,
+	storeKey storetypes.StoreKey,
+	tStoreKey storetypes.StoreKey,
+	authority string,
+	spKeeper types.SpKeeper,
+	accountKeeper types.AccountKeeper,
+	bankKeeper types.BankKeeper,
+	paymentKeeper types.PaymentKeeper,
+) *Keeper {
+
+	k := Keeper{
+		cdc:           cdc,
+		storeKey:      storeKey,
+		tStoreKey:     tStoreKey,
+		authority:     authority,
+		spKeeper:      spKeeper,
+		accountKeeper: accountKeeper,
+		bankKeeper:    bankKeeper,
+		paymentKeeper: paymentKeeper,
+	}
+
+	k.lvgSequence = sequence.NewSequence[uint32](types.LVGSequencePrefix)
+	k.gvgSequence = sequence.NewSequence[uint32](types.GVGSequencePrefix)
+	k.gvgFamilySequence = sequence.NewSequence[uint32](types.GVGFamilySequencePrefix)
+
+	return &k
+}
+
+func (k Keeper) GetAuthority() string {
+	return k.authority
+}
+
+func (k Keeper) Logger(ctx sdk.Context) log.Logger {
+	return ctx.Logger().With("module", fmt.Sprintf("x/%s", types.ModuleName))
+}
+
+func (k Keeper) GenNextLVGID(ctx sdk.Context) uint32 {
+	store := ctx.KVStore(k.storeKey)
+
+	seq := k.lvgSequence.NextVal(store)
+	return seq
+}
+
+func (k Keeper) GenNextGVGID(ctx sdk.Context) uint32 {
+	store := ctx.KVStore(k.storeKey)
+
+	seq := k.gvgSequence.NextVal(store)
+	return seq
+}
+
+func (k Keeper) GenNextGVGFamilyID(ctx sdk.Context) uint32 {
+	store := ctx.KVStore(k.storeKey)
+
+	seq := k.gvgFamilySequence.NextVal(store)
+	return seq
+}
+
+func (k Keeper) SetGVG(ctx sdk.Context, gvg *types.GlobalVirtualGroup) {
+	store := ctx.KVStore(k.storeKey)
+
+	bz := k.cdc.MustMarshal(gvg)
+	store.Set(types.GetGVGKey(gvg.Id), bz)
+}
+
+func (k Keeper) DeleteGVG(ctx sdk.Context, primarySpID, gvgID uint32) error {
+
+	store := ctx.KVStore(k.storeKey)
+
+	gvg, found := k.GetGVG(ctx, gvgID)
+	if !found {
+		return types.ErrGVGNotExist
+	}
+
+	if gvg.StoredSize != 0 {
```

Comment:
If empty object is stored in the gvg, it will still pass this check.

Reply:
Exactly, are we able to delete a GVG if it only serves for an empty GVG?